openssl-dynamic/src/main/java/io/netty/internal/tcnative/CertificateCallbackTask.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.internal.tcnative;
+
+/**
+ * Execute {@link CertificateCallback#handle(long, byte[], byte[][])}.
+ */
+final class CertificateCallbackTask extends SSLTask {
+    private final byte[] keyTypeBytes;
+    private final byte[][] asn1DerEncodedPrincipals;
+    private final CertificateCallback callback;
+
+    CertificateCallbackTask(long ssl, byte[] keyTypeBytes, byte[][] asn1DerEncodedPrincipals,
+                            CertificateCallback callback) {
+        super(ssl);
+        // It's ok to not clone the arrays as we create these in JNI and not-reuse.
+        this.keyTypeBytes = keyTypeBytes;
+        this.asn1DerEncodedPrincipals = asn1DerEncodedPrincipals;
+        this.callback = callback;
+    }
+
+    // See https://www.openssl.org/docs/man1.0.2/man3/SSL_set_cert_cb.html.
+    @Override
+    protected int runTask(long ssl) {
+        try {
+            callback.handle(ssl, keyTypeBytes, asn1DerEncodedPrincipals);
+            return 1;
+        } catch (Exception e) {
```

Comment:
It would be nice if this was logged at least.

Reply:
In netty we do before rethrow... we dont have any dependencies on a logging framework here. We could use java.util.logger ....

Reply:
That's what I was implying.  Else, you could just rethrow it as RuntimeException.

Reply:
@carl-mastrangelo the problem is that rethrowing here is really "useless" as we will process it as part of an openssl c callback which needs to return 0 for an error to abort the handshake.

Reply:
@carl-mastrangelo also I think we don't want to log as we rethrow it on the Netty level correctly after the `SSL.doHandshake(...)` method returns and so fail the handshake with an `SSLHandshakeException`.

Reply:
so to make it short I think it is all fine as it is... Maybe I should add some comments tho ?

Reply:
Comment is fine.